src/com/dabomstew/pkrandom/romhandlers/AbstractRomHandler.java

```diff
@@ -1925,6 +1924,76 @@ public void typeThemeTrainerPokes(Settings settings) {
         this.setTrainers(currentTrainers, false);
     }
 
+    @Override
+    public void randomizeTrainerHeldItems(Settings settings) {
+        boolean giveToBossPokemon = settings.isRandomizeHeldItemsForBossTrainerPokemon();
+        boolean giveToImportantPokemon = settings.isRandomizeHeldItemsForBossTrainerPokemon();
+        boolean giveToRegularPokemon = settings.isRandomizeHeldItemsForBossTrainerPokemon();
+        boolean highestLevelOnly = settings.isHighestLevelGetsItemsForTrainers();
+        List<Trainer> currentTrainers = this.getTrainers();
+        for (Trainer t : currentTrainers) {
+            if (!giveToRegularPokemon && (!t.isImportant() && !t.isBoss())) {
+                continue;
+            }
+            if (!giveToImportantPokemon && t.isImportant()) {
+                continue;
+            }
+            if (!giveToBossPokemon && t.isBoss()) {
+                continue;
+            }
+
+            if (highestLevelOnly) {
+                int maxLevel = -1;
+                TrainerPokemon highestLevelPoke = null;
+                for (TrainerPokemon tp : t.pokemon) {
+                    if (tp.level > maxLevel) {
+                        highestLevelPoke = tp;
+                        maxLevel = tp.level;
+                    }
+                    tp.heldItem = 0; // set every poke to no item (just in case there was one originally).

```

Comment:
I think it would be preferable to leave already existing items in place, rather than removing them (after all, the setting is "Add Held Items", not "Set Held Items").

Reply:
I agree. It also significantly reduces the complexity of handling this for generations 6 and 7, where the randomizer already does some special item handling to ensure Pokemon have the "proper" Mega Stone or Z-Crystal. If you kept the code like this, you would have to check whether one of those "special" items was being held before removing it.

Reply:
After talking with Ajarmar, I think my concern about Mega Stones and Z-Crystals can simply be solved by adding `hasMegaStone` and `hasZCrystal` properties to the TrainerPokemon class. The `getTrainers` method of the respective RomHandlers would set it based on the vanilla held item. Then, you can ignore held item randomization for Pokemon who have either of these properties set.

Reply:
I've changed it so we don't store a new item if there was already an item. I'm not quite sure I follow the Mega Stones and Z-Crystals discussion; in addition to not overwriting already-held items, do I need to check the ZCrystal/mega stone logic? Or does simply not writing a new item if there is one already cover those cases?

Reply:
Sorry, I think I ended up confusing you with mixed messages. This is what I think should happen, though Ajarmar can give his own ideas as to what should happen:
1. If a Pokemon has a Z-Crystal, don't change their item
2. If a Pokemon has a Mega Stone *and* "Swap Mega Evolvables" for trainer Pokemon is **true**, don't change their item
3. If "Highest Level Only" held item randomization is **true**, and Pokemon who are *not* the highest level have held items, don't change their item and keep it like it was
4. Otherwise, randomize their item

Does that make sense?